File: cpp-thin-client/src/thin_client.cpp
#include "ironwall/thin_client.hpp"
#include <iostream>
#include <thread>
#include <chrono>

namespace ironwall {

ThinClient::ThinClient(Config cfg, TeeAttestationResult att, ZkMovementValidator zk, DualAnchor anchors)
    : cfg_(std::move(cfg)), att_(std::move(att)), zk_(std::move(zk)), anchors_(std::move(anchors)) {}

void ThinClient::run() {
    std::cout << "[client] Thin client main loop started\n";
    std::cout << "[client] Attestation quote: " << att_.quote_hash << "\n";

    uint32_t tick_ms = 1000 / std::max(cfg_.tick_rate_hz, 1u);
    for (uint64_t tick = 0; tick < cfg_.demo_ticks; ++tick) {
        Vec3 from = position_;
        Vec3 to{from.x + 0.12f, from.y, from.z + 0.08f};

        try {
            auto proof = zk_.prove(cfg_.player_id, from, to, tick_ms);
            std::cout << "[client] tick=" << tick
                      << " pos=(" << to.x << "," << to.y << "," << to.z << ")"
                      << " speed=" << proof.speed
                      << " proof=" << proof.proof_id << "\n";

            auto receipt = anchors_.anchor(att_, proof);
            std::cout << "[client] anchored combined_hash=" << receipt.combined_hash << "\n";
            position_ = to;
        } catch (const InvalidMovement& e) {
            std::cout << "[client] movement rejected: " << e.what() << "\n";
        }

        std::this_thread::sleep_for(std::chrono::milliseconds(tick_ms));
    }
    std::cout << "[client] Demo loop finished\n";
}

} // namespace ironwall

File: rust-thin-client/src/thin_client.rs
use anyhow::Result;
use tracing::{info, warn};
use tokio::time::{sleep, Duration};

use crate::tee::TeeAttestationResult;
use crate::zk::ZkMovementValidator;
use crate::anchors::DualAnchor;
use crate::config::IronwallConfig;
use crate::error::IronwallError;

pub struct ThinClient {
    pub cfg: IronwallConfig,
    pub attestation: TeeAttestationResult,
    pub zk: ZkMovementValidator,
    pub anchors: DualAnchor,
    position: (f32, f32, f32),
}

impl ThinClient {
    pub fn new(
        cfg: IronwallConfig,
        attestation: TeeAttestationResult,
        zk: ZkMovementValidator,
        anchors: DualAnchor,
    ) -> Self {
        Self {
            cfg,
            attestation,
            zk,
            anchors,
            position: (0.0, 0.0, 0.0),
        }
    }

    pub async fn run(&mut self) -> Result<()> {
        info!("Thin client main loop started");
        info!("Attestation quote: {}", self.attestation.quote_hash);

        let tick_ms = 1000 / self.cfg.tick_rate_hz.max(1);
        let mut tick: u64 = 0;

        while tick < 5 {
            let from = self.position;

            let to = (
                from.0 + 0.12,
                from.1,
                from.2 + 0.08,
            );

            match self.zk.prove_valid_movement(
                &self.cfg.player_id,
                from,
                to,
                tick_ms,
            ).await {
                Ok(proof) => {
                    info!(
                        "tick={} pos=({:.2},{:.2},{:.2}) speed={:.2} proof={}",
                        tick, to.0, to.1, to.2, proof.speed, proof.proof_id
                    );

                    let receipt = self.anchors.anchor_proof(&self.attestation, &proof).await?;
                    info!("anchored combined_hash={}", receipt.combined_hash);
                    self.position = to;
                }
                Err(IronwallError::InvalidMovement(msg)) => {
                    warn!("movement rejected: {}", msg);
                }
                Err(e) => return Err(e.into()),
            }

            tick += 1;
            sleep(Duration::from_millis(tick_ms as u64)).await;
        }

        info!("Demo loop finished");
        Ok(())
    }
}
